### PR TITLE
Feat/leap cosmos metamask snap

### PR DESCRIPTION
**Pull request by codingki**
image

- Restructure wallet files
- fix wallet connect approval timeout
- add leap cosmos metamask snap 

https://github.com/graz-sh/graz/assets/39829726/401346c7-48d5-4541-8bde-26aa13007c98

### CODE REVIEW

File: packages/graz/src/actions/wallet/metamask-snap/index.ts [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+ *   });
+ * } catch (error: Error) {
+ *   console.error(error.message);
+ * }
+ * ```
+ *
+ *
+ */
+export const getMetamaskSnap = (params?: GetMetamaskSnap): Wallet => {
+  const ethereum = (window as EthereumWindow).ethereum;
```
Comment (grikomsn) [inline packages/graz/src/actions/wallet/metamask-snap/index.ts]: Preferably typed in env.d.ts or declare in this file.

File: packages/graz/src/actions/wallet/metamask-snap/index.ts [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+  if (ethereum && params) {
+    const getSnaps = async (): Promise<GetSnapsResponse> => {
+      const res = (await ethereum.request({
+        method: "wallet_getSnaps",
+      })) as unknown as GetSnapsResponse;
+      return res;
+    };
```
Comment (grikomsn) [inline packages/graz/src/actions/wallet/metamask-snap/index.ts:48]: Also this one.

File: packages/graz/src/actions/wallet/metamask-snap/index.ts [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+          [params.id]: params.params || {},
+        },
+      });
+    };
+
+    const init = async () => {
+      const clientVersion = await ethereum.request({
+        method: "web3_clientVersion",
+      });
+      const isFlaskDetected = (clientVersion as string[])?.includes("flask");
```
Comment (grikomsn) [inline packages/graz/src/actions/wallet/metamask-snap/index.ts:72]: This too if possible.

File: packages/graz/src/actions/wallet/metamask-snap/index.ts [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+      return {
+        signature: signature.signature,
+        signed: {
+          ...signature.signed,
+          accountNumber: `${modifiedAccountNumber.toString()}`,
+          authInfoBytes: new Uint8Array(Object.values(signature.signed.authInfoBytes)),
+          bodyBytes: new Uint8Array(Object.values(signature.signed.bodyBytes)),
+        },
+      } as unknown as DirectSignResponse;
```
Comment (grikomsn) [inline packages/graz/src/actions/wallet/metamask-snap/index.ts:118]: 🤔

File: packages/graz/src/actions/wallet/metamask-snap/index.ts [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+            },
+          },
+        },
+      });
+      if (!res) throw new Error("No response from Metamask");
+
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-expect-error
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-unsafe-member-access
+      (res as Key).pubKey = Uint8Array.from(Object.values(res.pubkey));
```
Comment (grikomsn) [inline packages/graz/src/actions/wallet/metamask-snap/index.ts:160]: Especially this, might be prone to prop uncertainty.

**Review by grikomsn — APPROVED** [review]
LGTM with several comments regarding `window.ethereum`

File: packages/graz/src/actions/wallet/metamask-snap/index.ts (hunk 4 above)
Comment (codingki) [reply packages/graz/src/actions/wallet/metamask-snap/index.ts:118]: ethereum.request 🤷‍♂️

**Pull request merged**